[CI] SmokeTestMultiNodeClientYamlTestSuiteIT {create/60_refresh/refresh=wait_for} fails

The following test failed on CI (see [this build](https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+g1gc/2970/console)): 

org.elasticsearch.smoketest.SmokeTestMultiNodeClientYamlTestSuiteIT test {yaml=create/60_refresh/refresh=wait_for waits until changes are visible in search}

It doesn't reproduce locally and I don't know what happened. The trace shows that the wait_for option timed out but I have no clue why:

  ```
1> [2017-05-28T00:04:33,592][INFO ][o.e.s.SmokeTestMultiNodeClientYamlTestSuiteIT] [test {yaml=cat.aliases/10_basic/Column headers}]: after test
  1> [2017-05-28T00:04:33,598][INFO ][o.e.s.SmokeTestMultiNodeClientYamlTestSuiteIT] [test {yaml=create/60_refresh/refresh=wait_for waits until changes are visible in search}]: before test
  1> [2017-05-28T00:05:03,705][INFO ][o.e.s.SmokeTestMultiNodeClientYamlTestSuiteIT] [test {yaml=create/60_refresh/refresh=wait_for waits until changes are visible in search}]: after test
  1> [2017-05-28T00:05:03,709][INFO ][o.e.s.SmokeTestMultiNodeClientYamlTestSuiteIT] Stash dump on failure [{
  1>   "stash" : {
  1>     "body" : null
  1>   }
  1> }]
ERROR   30.1s | SmokeTestMultiNodeClientYamlTestSuiteIT.test {yaml=create/60_refresh/refresh=wait_for waits until changes are visible in search} <<< FAILURES!
   > Throwable #1: java.lang.RuntimeException: Failure at [create/60_refresh:67]: listener timeout after waiting for [30000] ms
   > 	at __randomizedtesting.SeedInfo.seed([C14AB41884F38AA8:491E8BC22A0FE750]:0)
   > 	at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.executeSection(ESClientYamlSuiteTestCase.java:346)
   > 	at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.test(ESClientYamlSuiteTestCase.java:328)
   > 	at java.lang.Thread.run(Thread.java:748)
   > Caused by: java.io.IOException: listener timeout after waiting for [30000] ms
   > 	at org.elasticsearch.client.RestClient$SyncResponseListener.get(RestClient.java:660)
   > 	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:219)
   > 	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:191)
   > 	at org.elasticsearch.test.rest.yaml.ClientYamlTestClient.callApi(ClientYamlTestClient.java:169)
   > 	at org.elasticsearch.test.rest.yaml.ClientYamlTestExecutionContext.callApiInternal(ClientYamlTestExecutionContext.java:157)
   > 	at org.elasticsearch.test.rest.yaml.ClientYamlTestExecutionContext.callApi(ClientYamlTestExecutionContext.java:89)
   > 	at org.elasticsearch.test.rest.yaml.section.DoSection.execute(DoSection.java:221)
   > 	at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.executeSection(ESClientYamlSuiteTestCase.java:344)
   > 	... 37 more
```

--- a/core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java
+++ b/core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java
@@ -20,6 +20,7 @@ package org.elasticsearch.action.support.replication;
 
 import org.apache.logging.log4j.Logger;
 import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.apache.lucene.store.AlreadyClosedException;
 import org.elasticsearch.ElasticsearchException;
 import org.elasticsearch.ExceptionsHelper;
 import org.elasticsearch.action.ActionListener;
@@ -187,6 +188,8 @@ public class ReplicationOperation<
                 successfulShards.incrementAndGet();
                 try {
                     primary.updateLocalCheckpointForShard(response.allocationId(), response.localCheckpoint());
+                } catch (final AlreadyClosedException e) {
+                    // okay, the index was deleted or this shard was never activated after a relocation; fallthrough and finish normally
                 } catch (final Exception e) {
                     final String message = String.format(Locale.ROOT, "primary failed updating local checkpoint for replica %s", shard);
                     primary.failShard(message, e);
@@ -327,7 +330,10 @@ public class ReplicationOperation<
         ShardRouting routingEntry();
 
         /**
-         * fail the primary, typically due to the fact that the operation has learned the primary has been demoted by the master
+         * Fail the primary shard.
+         *
+         * @param message   the failure message
+         * @param exception the exception that triggered the failure
          */
         void failShard(String message, Exception exception);
 

--- a/core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java
+++ b/core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java
@@ -191,9 +191,9 @@ public class ReplicationOperation<
                 } catch (final AlreadyClosedException e) {
                     // okay, the index was deleted or this shard was never activated after a relocation; fallthrough and finish normally
                 } catch (final Exception e) {
+                    // fail the primary but fall through and let the rest of operation processing complete
                     final String message = String.format(Locale.ROOT, "primary failed updating local checkpoint for replica %s", shard);
                     primary.failShard(message, e);
-                    finishAsFailed(new RetryOnPrimaryException(primary.routingEntry().shardId(), message, e));
                 }
                 decPendingAndFinishIfNeeded();
             }

--- a/core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java
+++ b/core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java
@@ -185,7 +185,13 @@ public class ReplicationOperation<
             @Override
             public void onResponse(ReplicaResponse response) {
                 successfulShards.incrementAndGet();
-                primary.updateLocalCheckpointForShard(response.allocationId(), response.localCheckpoint());
+                try {
+                    primary.updateLocalCheckpointForShard(response.allocationId(), response.localCheckpoint());
+                } catch (final Exception e) {
+                    final String message = String.format(Locale.ROOT, "primary failed updating local checkpoint for replica %s", shard);
+                    primary.failShard(message, e);
+                    finishAsFailed(new RetryOnPrimaryException(primary.routingEntry().shardId(), message, e));
+                }
                 decPendingAndFinishIfNeeded();
             }
 
@@ -334,7 +340,6 @@ public class ReplicationOperation<
          * @return the request to send to the replicas
          */
         PrimaryResultT perform(RequestT request) throws Exception;
-
 
         /**
          * Notifies the primary of a local checkpoint for the given allocation.

--- a/core/src/test/java/org/elasticsearch/action/support/replication/ReplicationOperationTests.java
+++ b/core/src/test/java/org/elasticsearch/action/support/replication/ReplicationOperationTests.java
@@ -343,17 +343,14 @@ public class ReplicationOperationTests extends ESTestCase {
 
         if (fatal) {
             assertTrue(primaryFailed.get());
-            final ExecutionException e = expectThrows(ExecutionException.class, listener::get);
-            assertThat(e.getCause(), instanceOf(ReplicationOperation.RetryOnPrimaryException.class));
-            final ReplicationOperation.RetryOnPrimaryException rope = (ReplicationOperation.RetryOnPrimaryException) e.getCause();
-            assertThat(rope, hasToString(containsString("primary failed updating local checkpoint for replica")));
         } else {
             assertFalse(primaryFailed.get());
-            final ShardInfo shardInfo = listener.actionGet().getShardInfo();
-            assertThat(shardInfo.getFailed(), equalTo(0));
-            assertThat(shardInfo.getFailures(), arrayWithSize(0));
-            assertThat(shardInfo.getSuccessful(), equalTo(1 + getExpectedReplicas(shardId, state).size()));
         }
+        assertThat(primaryFailed.get(), equalTo(fatal));
+        final ShardInfo shardInfo = listener.actionGet().getShardInfo();
+        assertThat(shardInfo.getFailed(), equalTo(0));
+        assertThat(shardInfo.getFailures(), arrayWithSize(0));
+        assertThat(shardInfo.getSuccessful(), equalTo(1 + getExpectedReplicas(shardId, state).size()));
     }
 
     private Set<ShardRouting> getExpectedReplicas(ShardId shardId, ClusterState state) {

--- a/core/src/test/java/org/elasticsearch/action/support/replication/ReplicationOperationTests.java
+++ b/core/src/test/java/org/elasticsearch/action/support/replication/ReplicationOperationTests.java
@@ -20,6 +20,7 @@ package org.elasticsearch.action.support.replication;
 
 import org.apache.logging.log4j.Logger;
 import org.apache.lucene.index.CorruptIndexException;
+import org.apache.lucene.store.AlreadyClosedException;
 import org.elasticsearch.ElasticsearchException;
 import org.elasticsearch.action.ActionListener;
 import org.elasticsearch.action.UnavailableShardsException;
@@ -56,7 +57,9 @@ import java.util.function.Supplier;
 import static org.elasticsearch.action.support.replication.ClusterStateCreationUtils.state;
 import static org.elasticsearch.action.support.replication.ClusterStateCreationUtils.stateWithActivePrimary;
 import static org.hamcrest.Matchers.arrayWithSize;
+import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasToString;
 import static org.hamcrest.Matchers.instanceOf;
 import static org.hamcrest.Matchers.notNullValue;
 import static org.hamcrest.Matchers.nullValue;
@@ -191,8 +194,7 @@ public class ReplicationOperationTests extends ESTestCase {
                 assertTrue(primaryFailed.compareAndSet(false, true));
             }
         };
-        final TestReplicationOperation op = new TestReplicationOperation(request, primary, listener, replicasProxy,
-            () -> finalState);
+        final TestReplicationOperation op = new TestReplicationOperation(request, primary, listener, replicasProxy, () -> finalState);
         op.execute();
 
         assertThat("request was not processed on primary", request.processedOnPrimary.get(), equalTo(true));
@@ -297,6 +299,49 @@ public class ReplicationOperationTests extends ESTestCase {
                 request.processedOnPrimary.get());
             assertListenerThrows("should throw exception to trigger retry", listener, UnavailableShardsException.class);
         }
+    }
+
+    public void testPrimaryFailureHandlingReplicaResponse() throws Exception {
+        final String index = "test";
+        final ShardId shardId = new ShardId(index, "_na_", 0);
+
+        final Request request = new Request(shardId);
+
+        final ClusterState state = stateWithActivePrimary(index, true, 1, 0);
+        final IndexMetaData indexMetaData = state.getMetaData().index(index);
+        final long primaryTerm = indexMetaData.primaryTerm(0);
+        final ShardRouting primaryRouting = state.getRoutingTable().shardRoutingTable(shardId).primaryShard();
+
+        final AtomicBoolean primaryFailed = new AtomicBoolean();
+        final ReplicationOperation.Primary<Request, Request, TestPrimary.Result> primary = new TestPrimary(primaryRouting, primaryTerm) {
+
+            @Override
+            public void failShard(String message, Exception exception) {
+                primaryFailed.set(true);
+            }
+
+            @Override
+            public void updateLocalCheckpointForShard(String allocationId, long checkpoint) {
+                if (primaryRouting.allocationId().getId().equals(allocationId)) {
+                    super.updateLocalCheckpointForShard(allocationId, checkpoint);
+                } else {
+                    throw new AlreadyClosedException("closed");
+                }
+            }
+
+        };
+
+        final PlainActionFuture<TestPrimary.Result> listener = new PlainActionFuture<>();
+        final ReplicationOperation.Replicas<Request> replicas = new TestReplicaProxy(Collections.emptyMap());
+        TestReplicationOperation operation = new TestReplicationOperation(request, primary, listener, replicas, () -> state);
+        operation.execute();
+
+        assertTrue(primaryFailed.get());
+        final ExecutionException e = expectThrows(ExecutionException.class, listener::get);
+        assertThat(e.getCause(), instanceOf(ReplicationOperation.RetryOnPrimaryException.class));
+        final ReplicationOperation.RetryOnPrimaryException rope = (ReplicationOperation.RetryOnPrimaryException) e.getCause();
+        assertThat(rope, hasToString(containsString("primary failed updating local checkpoint for replica")));
+        // no need to assert that no replica was failed, the test replica proxy already ensures this since we passed an empty failure map
     }
 
     private Set<ShardRouting> getExpectedReplicas(ShardId shardId, ClusterState state) {

--- a/core/src/test/java/org/elasticsearch/action/support/replication/ReplicationOperationTests.java
+++ b/core/src/test/java/org/elasticsearch/action/support/replication/ReplicationOperationTests.java
@@ -312,6 +312,7 @@ public class ReplicationOperationTests extends ESTestCase {
         final long primaryTerm = indexMetaData.primaryTerm(0);
         final ShardRouting primaryRouting = state.getRoutingTable().shardRoutingTable(shardId).primaryShard();
 
+        final boolean fatal = randomBoolean();
         final AtomicBoolean primaryFailed = new AtomicBoolean();
         final ReplicationOperation.Primary<Request, Request, TestPrimary.Result> primary = new TestPrimary(primaryRouting, primaryTerm) {
 
@@ -325,7 +326,11 @@ public class ReplicationOperationTests extends ESTestCase {
                 if (primaryRouting.allocationId().getId().equals(allocationId)) {
                     super.updateLocalCheckpointForShard(allocationId, checkpoint);
                 } else {
-                    throw new AlreadyClosedException("closed");
+                    if (fatal) {
+                        throw new NullPointerException();
+                    } else {
+                        throw new AlreadyClosedException("already closed");
+                    }
                 }
             }
 
@@ -336,12 +341,19 @@ public class ReplicationOperationTests extends ESTestCase {
         TestReplicationOperation operation = new TestReplicationOperation(request, primary, listener, replicas, () -> state);
         operation.execute();
 
-        assertTrue(primaryFailed.get());
-        final ExecutionException e = expectThrows(ExecutionException.class, listener::get);
-        assertThat(e.getCause(), instanceOf(ReplicationOperation.RetryOnPrimaryException.class));
-        final ReplicationOperation.RetryOnPrimaryException rope = (ReplicationOperation.RetryOnPrimaryException) e.getCause();
-        assertThat(rope, hasToString(containsString("primary failed updating local checkpoint for replica")));
-        // no need to assert that no replica was failed, the test replica proxy already ensures this since we passed an empty failure map
+        if (fatal) {
+            assertTrue(primaryFailed.get());
+            final ExecutionException e = expectThrows(ExecutionException.class, listener::get);
+            assertThat(e.getCause(), instanceOf(ReplicationOperation.RetryOnPrimaryException.class));
+            final ReplicationOperation.RetryOnPrimaryException rope = (ReplicationOperation.RetryOnPrimaryException) e.getCause();
+            assertThat(rope, hasToString(containsString("primary failed updating local checkpoint for replica")));
+        } else {
+            assertFalse(primaryFailed.get());
+            final ShardInfo shardInfo = listener.actionGet().getShardInfo();
+            assertThat(shardInfo.getFailed(), equalTo(0));
+            assertThat(shardInfo.getFailures(), arrayWithSize(0));
+            assertThat(shardInfo.getSuccessful(), equalTo(1 + getExpectedReplicas(shardId, state).size()));
+        }
     }
 
     private Set<ShardRouting> getExpectedReplicas(ShardId shardId, ClusterState state) {

--- a/core/src/test/java/org/elasticsearch/action/support/replication/ReplicationOperationTests.java
+++ b/core/src/test/java/org/elasticsearch/action/support/replication/ReplicationOperationTests.java
@@ -341,11 +341,6 @@ public class ReplicationOperationTests extends ESTestCase {
         TestReplicationOperation operation = new TestReplicationOperation(request, primary, listener, replicas, () -> state);
         operation.execute();
 
-        if (fatal) {
-            assertTrue(primaryFailed.get());
-        } else {
-            assertFalse(primaryFailed.get());
-        }
         assertThat(primaryFailed.get(), equalTo(fatal));
         final ShardInfo shardInfo = listener.actionGet().getShardInfo();
         assertThat(shardInfo.getFailed(), equalTo(0));

